### PR TITLE
fix: ingress rules order is preserved

### LINKED ISSUE #91: Flaky tests at `Kong` ToGateway
**What happened**:
Run unit tests at the main branch.

**What you expected to happen**:
All tests should pass

**How to reproduce it (as minimally and precisely as possible)**:
`go test -v -race ./... -count=10`

**Anything else we need to know?**:
The failed tests passes most of the time, however in some cases it won't

Error log:
```
=== RUN   Test_ToGateway/header_matching,_method_matching,_multiple_ingress_rules
    converter_test.go:348: Expected HTTPRoute default/multiple-matching-multiple-rules-test-mydomain-com to be {TypeMeta:{Kind:HTTPRoute APIVersion:gateway.networking.k8s.io/v1beta1} ObjectMeta:{Name:multiple-matching-multiple-rules-test-mydomain-com GenerateName: Namespace:default SelfLink: UID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ManagedFields:[]} Spec:{CommonRouteSpec:{ParentRefs:[{Group:<nil> Kind:<nil> Namespace:<nil> Name:ingress-kong SectionName:<nil> Port:<nil>}]} Hostnames:[test.mydomain.com] Rules:[{Matches:[{Path:0xc000143910 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc000143890} {Path:0xc000143920 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc0001438b0}] Filters:[] BackendRefs:[{BackendRef:{BackendObjectReference:{Group:<nil> Kind:<nil> Name:test-first Namespace:<nil> Port:0xc0004a11cc} Weight:<nil>} Filters:[]}]} {Matches:[{Path:0xc000143930 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc0001438d0} {Path:0xc000143940 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc0001438f0}] Filters:[] BackendRefs:[{BackendRef:{BackendObjectReference:{Group:<nil> Kind:<nil> Name:test-second Namespace:<nil> Port:0xc0004a11dc} Weight:<nil>} Filters:[]}]}]} Status:{RouteStatus:{Parents:[]}}}
         Got: {TypeMeta:{Kind:HTTPRoute APIVersion:gateway.networking.k8s.io/v1beta1} ObjectMeta:{Name:multiple-matching-multiple-rules-test-mydomain-com GenerateName: Namespace:default SelfLink: UID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ManagedFields:[]} Spec:{CommonRouteSpec:{ParentRefs:[{Group:<nil> Kind:<nil> Namespace:<nil> Name:ingress-kong SectionName:<nil> Port:<nil>}]} Hostnames:[test.mydomain.com] Rules:[{Matches:[{Path:0xc000244500 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc0002444f0} {Path:0xc000244540 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc000244530}] Filters:[] BackendRefs:[{BackendRef:{BackendObjectReference:{Group:<nil> Kind:<nil> Name:test-second Namespace:<nil> Port:0xc000002cb0} Weight:<nil>} Filters:[]}]} {Matches:[{Path:0xc000244580 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc000244570} {Path:0xc0002445c0 Headers:[{Type:<nil> Name:key1 Value:val1}] QueryParams:[] Method:0xc0002445b0}] Filters:[] BackendRefs:[{BackendRef:{BackendObjectReference:{Group:<nil> Kind:<nil> Name:test-first Namespace:<nil> Port:0xc000002c80} Weight:<nil>} Filters:[]}]}]} Status:{RouteStatus:{Parents:[]}}}
         Diff:   v1beta1.HTTPRoute{
          	TypeMeta:   {Kind: "HTTPRoute", APIVersion: "gateway.networking.k8s.io/v1beta1"},
          	ObjectMeta: {Name: "multiple-matching-multiple-rules-test-mydomain-com", Namespace: "default"},
          	Spec: v1beta1.HTTPRouteSpec{
          		CommonRouteSpec: {ParentRefs: {{Name: "ingress-kong"}}},
          		Hostnames:       {"test.mydomain.com"},
          		Rules: []v1beta1.HTTPRouteRule{
          			{
          				Matches: []v1beta1.HTTPRouteMatch{
          					{
          						Path: &v1beta1.HTTPPathMatch{
          							Type:  &"PathPrefix",
        - 							Value: &"/first",
        + 							Value: &"/second",
          						},
          						Headers:     {{Name: "key1", Value: "val1"}},
          						QueryParams: nil,
          						Method:      &"GET",
          					},
          					{
          						Path: &v1beta1.HTTPPathMatch{
          							Type:  &"PathPrefix",
        - 							Value: &"/first",
        + 							Value: &"/second",
          						},
          						Headers:     {{Name: "key1", Value: "val1"}},
          						QueryParams: nil,
          						Method:      &"POST",
          					},
          				},
        - 				Filters: nil,
        + 				Filters: []v1beta1.HTTPRouteFilter{},
          				BackendRefs: []v1beta1.HTTPBackendRef{
          					{
          						BackendRef: v1beta1.BackendRef{
          							BackendObjectReference: v1beta1.BackendObjectReference{
          								Group:     nil,
          								Kind:      nil,
        - 								Name:      "test-first",
        + 								Name:      "test-second",
          								Namespace: nil,
          								Port:      &80,
          							},
          							Weight: nil,
          						},
          						Filters: nil,
          					},
          				},
          			},
          			{
          				Matches: []v1beta1.HTTPRouteMatch{
          					{
          						Path: &v1beta1.HTTPPathMatch{
          							Type:  &"PathPrefix",
        - 							Value: &"/second",
        + 							Value: &"/first",
          						},
          						Headers:     {{Name: "key1", Value: "val1"}},
          						QueryParams: nil,
          						Method:      &"GET",
          					},
          					{
          						Path: &v1beta1.HTTPPathMatch{
          							Type:  &"PathPrefix",
        - 							Value: &"/second",
        + 							Value: &"/first",
          						},
          						Headers:     {{Name: "key1", Value: "val1"}},
          						QueryParams: nil,
          						Method:      &"POST",
          					},
          				},
        - 				Filters: nil,
        + 				Filters: []v1beta1.HTTPRouteFilter{},
          				BackendRefs: []v1beta1.HTTPBackendRef{
          					{
          						BackendRef: v1beta1.BackendRef{
          							BackendObjectReference: v1beta1.BackendObjectReference{
          								Group:     nil,
          								Kind:      nil,
        - 								Name:      "test-second",
        + 								Name:      "test-first",
          								Namespace: nil,
          								Port:      &80,
          							},
          							Weight: nil,
          						},
          						Filters: nil,
          					},
          				},
          			},
          		},
          	},
        - 	Status: v1beta1.HTTPRouteStatus{},
        + 	Status: v1beta1.HTTPRouteStatus{RouteStatus: v1beta1.RouteStatus{Parents: []v1beta1.RouteParentStatus{}}},
          }
--- FAIL: Test_ToGateway (0.01s)
    --- PASS: Test_ToGateway/header_matching,_method_matching,_plugin,_single_ingress_rule (0.00s)
    --- FAIL: Test_ToGateway/header_matching,_method_matching,_multiple_ingress_rules (0.01s)
```

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Ingress Rules order needs to be preserved when converting into `HTTPRoute` rules.

**Which issue(s) this PR fixes**:

Fixes #91

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by LiorLieberman** [comment]
@mlavacca I think we want to make sure we dont produce different rule orders other than making the tests pass with different order. 
Changing the rule orders have implications.  Also it can result in unnecessary update to the API(in case the HTTPRoute is already in the cluster)

**Comment by mlavacca** [comment]
> @mlavacca I think we want to make sure we dont produce different rule orders other than making the tests pass with different order. Changing the rule orders have implications. Also it can result in unnecessary update to the API(in case the HTTPRoute is already in the cluster)

yep, I agree that we should preserve the rule orders. I dug a bit into the issue, and the problem is caused by the fact that in the common [package we aggregate the ingress rules into a map](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/common/converter.go#L272), and then we iterate that map to create HTTPRoute rules. Since map order is not guaranteed, this can cause flakes. This problem has been surfaced by the new kong test as there were no tests with ingresses having multiple rules previously. We may need to add some logic to fix it. I'll do it in this PR.

File: pkg/i2gw/providers/common/converter.go [hunk 1]
```diff
@@ -272,10 +272,16 @@ func (rg *ingressRuleGroup) toHTTPRoute(options i2gw.ProviderImplementationSpeci
...
+	uniqueKeys := map[int]string{}
+	var index int
 	for i, ir := range rg.rules {
 		for j, path := range ir.rule.HTTP.Paths {
 			ip := ingressPath{ruleIdx: i, pathIdx: j, ruleType: "http", path: path}
 			pmKey := getPathMatchKey(ip)
+			if _, ok := pathsByMatchGroup[pmKey]; !ok && len(pathsByMatchGroup) > 0 {
+				index++
+			}
+			uniqueKeys[index] = string(pmKey)
```
Comment (levikobi) [inline pkg/i2gw/providers/common/converter.go:284]: I'm not sure if it is possible, but imagine the following scenario:
1. pmKey = x -> (!ok && len(pathsByMatchGroup) > 0) == false -> index 0 -> uniqueKeys[0] points to pmKey x
2. pmKey = y -> (!ok && len(pathsByMatchGroup) > 0) == true -> index 1 -> uniqueKeys[1] points to pmKey y
3. pmKey = x -> (!ok && len(pathsByMatchGroup) > 0) == false -> index 1 -> uniqueKeys[1] now points to pmKey x instead of y.

Although I may be missing out on something here.

Also, could you please add a small comment here? I find this code block a bit confusing (or at least not straightforward)

**Review by levikobi — COMMENTED** [review]
Thanks for tackling this one @mlavacca!

File: pkg/i2gw/providers/common/converter.go (hunk 1 above)
Comment (mlavacca) [reply pkg/i2gw/providers/common/converter.go:284]: You are right, nice catch! I'll extract this part into a function and create tests for it. There may be corner cases and I want a proper test coverage.

Comment (levikobi) [reply pkg/i2gw/providers/common/converter.go:284]: Thanks 😅
That sounds good to me. Unit tests will be great here.

Comment (mlavacca) [reply pkg/i2gw/providers/common/converter.go]: function changed and tests added

**Review by LiorLieberman — COMMENTED** [review]
Throwing an idea. 
Is this helping also achieving what we want? 

```go
type orderedPathsByMatchGroup struct {
	keys []pathMatchKey
	data map[pathMatchKey][]ingressPath
}

func (rg *ingressRuleGroup) toHTTPRoute() (gatewayv1beta1.HTTPRoute, field.ErrorList) {
	// pathsByMatchGroup := map[pathMatchKey][]ingressPath{}
	pathsByMatchGroup := &orderedPathsByMatchGroup{
		keys: []pathMatchKey{},
		data: make(map[pathMatchKey][]ingressPath),
	}
	var errors field.ErrorList

	for i, ir := range rg.rules {
		for j, path := range ir.rule.HTTP.Paths {
			ip := ingressPath{ruleIdx: i, pathIdx: j, ruleType: "http", path: path}
			pmKey := getPathMatchKey(ip)
			if _, ok := pathsByMatchGroup.data[pmKey]; !ok {
				pathsByMatchGroup.keys = append(pathsByMatchGroup.keys, pmKey)
			}
			pathsByMatchGroup.data[pmKey] = append(pathsByMatchGroup.data[pmKey], ip)

		}
	}
```

later in the range:

```go
	for _, pmKey := range pathsByMatchGroup.keys {
		paths := pathsByMatchGroup.data[pmKey]
		path := paths[0]
```

**Comment by mlavacca** [comment]
> Throwing an idea. Is this helping also achieving what we want?
> 
> ```go
> type orderedPathsByMatchGroup struct {
> 	keys []pathMatchKey
> 	data map[pathMatchKey][]ingressPath
> }
> 
> func (rg *ingressRuleGroup) toHTTPRoute() (gatewayv1beta1.HTTPRoute, field.ErrorList) {
> 	// pathsByMatchGroup := map[pathMatchKey][]ingressPath{}
> 	pathsByMatchGroup := &orderedPathsByMatchGroup{
> 		keys: []pathMatchKey{},
> 		data: make(map[pathMatchKey][]ingressPath),
> 	}
> 	var errors field.ErrorList
> 
> 	for i, ir := range rg.rules {
> 		for j, path := range ir.rule.HTTP.Paths {
> 			ip := ingressPath{ruleIdx: i, pathIdx: j, ruleType: "http", path: path}
> 			pmKey := getPathMatchKey(ip)
> 			if _, ok := pathsByMatchGroup.data[pmKey]; !ok {
> 				pathsByMatchGroup.keys = append(pathsByMatchGroup.keys, pmKey)
> 			}
> 			pathsByMatchGroup.data[pmKey] = append(pathsByMatchGroup.data[pmKey], ip)
> 
> 		}
> 	}
> ```
> 
> later in the range:
> 
> ```go
> 	for _, pmKey := range pathsByMatchGroup.keys {
> 		paths := pathsByMatchGroup.data[pmKey]
> 		path := paths[0]
> ```
> 
> By keeping only the keys in a slice, we can also retain the order in which the elements were added. I have not ran this but it feels like it could work

It makes sense, your proposal is more efficient. I've made this change and changed the tests accordingly.

File: pkg/i2gw/providers/common/utils.go [hunk 2]
```diff
@@ -127,6 +127,34 @@ func ToBackendRef(ib networkingv1.IngressBackend, path *field.Path) (*gatewayv1b
 	}, nil
 }
 
+type orderedPathsByMatchGroup struct {
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils.go:130]:
```suggestion
type orderedIngressPathsByMatchKey struct {
```

File: pkg/i2gw/providers/common/utils.go [hunk 3]
```diff
@@ -127,6 +127,34 @@ func ToBackendRef(ib networkingv1.IngressBackend, path *field.Path) (*gatewayv1b
...
+	keys []pathMatchKey
+	data map[pathMatchKey][]ingressPath
+}
+
+func groupPaths(rules []ingressRule) orderedPathsByMatchGroup {
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils.go:135]:
```suggestion
func groupIngressPathsByMatchKey(rules []ingressRule) orderedIngressPathsByMatchKey {
```

File: pkg/i2gw/providers/common/utils.go [hunk 4]
```diff
@@ -127,6 +127,34 @@ func ToBackendRef(ib networkingv1.IngressBackend, path *field.Path) (*gatewayv1b
...
+	pathsByMatchGroup := orderedPathsByMatchGroup{}
+
+	for i, ir := range rules {
+		for j, path := range ir.rule.HTTP.Paths {
+			if len(pathsByMatchGroup.keys) == 0 {
+				pathsByMatchGroup.keys = []pathMatchKey{}
+			}
+			if len(pathsByMatchGroup.data) == 0 {
+				pathsByMatchGroup.data = map[pathMatchKey][]ingressPath{}
+			}
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils.go:146]:
```suggestion
	ingressPathsByMatchKey := & orderedIngressPathsByMatchKey{
		keys: []pathMatchKey{},
		data: make(map[pathMatchKey][]ingressPath),
	}

	for i, ir := range rules {
		for j, path := range ir.rule.HTTP.Paths {
```

File: pkg/i2gw/providers/common/converter.go
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/converter.go:1]: s/pathsByMatchGroup/ingressPathsByMatchKey/g

**Review by LiorLieberman — COMMENTED** [review]
Thanks Mattia!

File: pkg/i2gw/providers/common/utils.go (hunk 4 above)
Comment (mlavacca) [reply pkg/i2gw/providers/common/utils.go]: I considered it, but I think that having structures to nil instead of being empty is a better pattern in case no rule exists. From a functional point of view, the approaches are equivalent.

Comment (LiorLieberman) [reply pkg/i2gw/providers/common/utils.go]: why a better pattern? I think it makes the code less readable and adds additional two if statements.  
memory wise, it is not really an optimization and it is anyway cleaned when the function finishes.

Comment (mlavacca) [reply pkg/i2gw/providers/common/utils.go]: I think this is the field of personal opinions. I prefer one way over another, but there is no difference functionally and performance-wise. If you prefer that way I'm ok with changing it.

File: pkg/i2gw/providers/common/utils.go [hunk 5]
```diff
@@ -127,6 +127,31 @@ func ToBackendRef(ib networkingv1.IngressBackend, path *field.Path) (*gatewayv1b
 	}, nil
 }
 
+type orderedIngressPathsByMatchKey struct {
+	keys []pathMatchKey
+	data map[pathMatchKey][]ingressPath
+}
+
+func groupIngressPathsByMatchKey(rules []ingressRule) orderedIngressPathsByMatchKey {
+	// we use a slice instead of a map to preserve rules order
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils.go:136]:
```suggestion
	// we track the keys in a slice to preserve the rules order
```

File: pkg/i2gw/providers/common/converter.go [hunk 6]
```diff
@@ -269,17 +269,7 @@ func (a *ingressAggregator) toHTTPRoutesAndGateways(options i2gw.ProviderImpleme
...
-
-	for i, ir := range rg.rules {
-		for j, path := range ir.rule.HTTP.Paths {
-			ip := ingressPath{ruleIdx: i, pathIdx: j, ruleType: "http", path: path}
-			pmKey := getPathMatchKey(ip)
-			pathsByMatchGroup[pmKey] = append(pathsByMatchGroup[pmKey], ip)
-		}
-	}
-
+	pathsByMatchGroup := groupIngressPathsByMatchKey(rg.rules)
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/converter.go:272]:
```suggestion
	ingressPathsByMatchKey := groupIngressPathsByMatchKey(rg.rules)
```

File: pkg/i2gw/providers/common/converter.go [hunk 7]
```diff
@@ -301,7 +291,9 @@ func (rg *ingressRuleGroup) toHTTPRoute(options i2gw.ProviderImplementationSpeci
 		httpRoute.Spec.Hostnames = []gatewayv1beta1.Hostname{gatewayv1beta1.Hostname(rg.host)}
 	}
 
-	for _, paths := range pathsByMatchGroup {
+	var errors field.ErrorList
+	for _, key := range pathsByMatchGroup.keys {
+		paths := pathsByMatchGroup.data[key]
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/converter.go:296]:
```suggestion
	for _, key := range ingressPathsByMatchKey.keys {
		paths := ingressPathsByMatchKey.data[key]
```

**Review by LiorLieberman — APPROVED** [review]

File: pkg/i2gw/providers/common/utils_test.go [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+package common
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	networkingv1 "k8s.io/api/networking/v1"
+)
+
+func TestGroupPaths(t *testing.T) {
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/common/utils_test.go:26]:
```suggestion
func TestGroupIngressPathsByMatchKey(t *testing.T) {
```

**Comment by LiorLieberman** [comment]
Thanks Mattia!
/lgtm
